src/main/java/org/logstash/tcp/InputLoop.java

```diff
@@ -0,0 +1,198 @@
+package org.logstash.tcp;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.StandardSocketOptions;
+import java.nio.ByteBuffer;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.channels.SelectionKey;
+import java.nio.channels.Selector;
+import java.nio.channels.ServerSocketChannel;
+import java.nio.channels.SocketChannel;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+public final class InputLoop implements Runnable, Closeable {
+
+    /**
+     * Read buffer size.
+     */
+    private static final int BUFFER_SIZE = 8096;
```

Comment:
so max payload is 8k, else it gets truncated ? 

Seems like a reasonable upperbound for most of our use cases (as I understand them), perhaps we should wrap this in java system.property with 8k as the default for the unexpected use case ?